chain-signatures/node/src/checkpoint.rs

```diff
@@ -0,0 +1,432 @@
+            processed_block_height: None,
+        }
+    }
+
+    /// Inserts a sign-respond transaction into the pending requests map
+    /// Returns Some(old_value) if the key was already present
+    pub fn insert(&mut self, id: SignId, tx: SignRespondTx) -> Option<SignRespondTx> {
+        self.requests.insert(id, tx)
+    }
+
+    /// Removes a sign-respond transaction from the pending requests map
+    /// Returns Some(value) if the key was present
+    pub fn remove(&mut self, id: &SignId) -> Option<SignRespondTx> {
+        self.requests.remove(id)
+    }
+
+    /// Gets a clone of a sign-respond transaction from the pending requests map
+    /// Returns Some(value) if the key is present
+    pub fn get(&self, id: &SignId) -> Option<SignRespondTx> {
+        self.requests.get(id).cloned()
+    }
+
+    /// Returns the number of pending requests
+    pub fn len(&self) -> usize {
+        self.requests.len()
+    }
+
+    /// Returns true if there are no pending requests
+    pub fn is_empty(&self) -> bool {
+        self.requests.is_empty()
+    }
+
+    /// Returns all sign-respond transactions with a specific status
+    pub fn get_by_status(&self, status: PendingRequestStatus) -> HashMap<SignId, SignRespondTx> {
+        self.requests
+            .iter()
+            .filter(|(_, tx)| tx.status == status)
+            .map(|(id, tx)| (*id, tx.clone()))
+            .collect()
+    }
+
+    /// Add a transaction ID to the pending publish queue
+    pub fn push_pending_publish(&mut self, id: SignId) {
+        self.pending_publish.push_back(id);
+    }
+
+    /// Pop the next transaction ID from the pending publish queue
+    /// Returns the transaction if found
+    pub fn pop_pending_publish(&mut self) -> Option<(SignId, SignRespondTx)> {
+        let id = self.pending_publish.pop_front()?;
+        let tx = self.get(&id)?;
+        Some((id, tx))
+    }
+
+    /// Get the number of transactions waiting to be published
+    pub fn pending_publish_count(&self) -> usize {
+        self.pending_publish.len()
+    }
+
+    /// Get all pending transaction IDs and their transactions
+    pub fn get_pending_execution(&self) -> Vec<(SignId, SignRespondTx)> {
+        self.pending_publish
+            .iter()
+            .filter_map(|id| self.get(id).map(|tx| (*id, tx)))
+            .collect()
+    }
+
+    /// Get the processed block height for this chain
+    pub fn processed_block_height(&self) -> Option<u64> {
+        self.processed_block_height
+    }
+
+    /// Set the processed block height for this chain
+    pub fn set_processed_block_height(&mut self, height: u64) {
+        self.processed_block_height = Some(height);
+    }
+}
+
+/// Checkpoints manages pending sign-respond requests across multiple chains
+/// Each chain has its own isolated set of pending requests with their own publish queues
+#[derive(Debug, Clone)]
+pub struct Checkpoints {
+    requests: Arc<RwLock<HashMap<Chain, PendingRequests>>>,
```

Comment:
note: To avoid lock contention, we might want to consider `requests: Arc<HashMap<Chain, RwLock<PendingRequests>>>` instead.

But let's avoid premature optimization. Just something to keep in mind, as we are building this to be a central piece used by many concurrent tasks.